### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [1]:
# Dependencies and Setup
import pandas as pd

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)

## Player Count

* Display the total number of players


In [2]:
# Create a variable that shows the number of total players
total_players = purchase_data["SN"].nunique()

# Create a summary table to show the total number of players
total_players_st = pd.DataFrame([{"Number of Total Players": total_players}])
total_players_st.head()

Number of Total Players
0                      576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [3]:
# Find and store the number of unique items using item name
unique_item_count = purchase_data["Item Name"].nunique()

# Find and store the average purchase price
average_purchase_price = purchase_data["Price"].mean()

# Find and store the total number of purchases by using Purchase ID
number_of_purchases = purchase_data["Purchase ID"].count()

# Find and store the total revenue
total_revenue = purchase_data["Price"].sum()

# Create the analysis data in a summary table
purchase_analysis_st = pd.DataFrame([{"Number of Unique Items":unique_item_count,
                                      "Average Purchase Price":average_purchase_price,
                                     "Total Number of Purchases":number_of_purchases,
                                     "Total Revenue":total_revenue}])

# Format two columns as a price
purchase_analysis_st["Average Purchase Price"] = purchase_analysis_st["Average Purchase Price"].map("${:,.2f}".format)
purchase_analysis_st["Total Revenue"] = purchase_analysis_st["Total Revenue"].map("${:,.2f}".format)

# Display the summary table
purchase_analysis_st.head()

Number of Unique Items Average Purchase Price  Total Number of Purchases  \
0                     179                  $3.05                        780   

  Total Revenue  
0     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [4]:
# Create a dataframe with purchase data for male players
male_purchase_data = purchase_data.loc[purchase_data["Gender"] == "Male",:]

# Find the count and percentage of male players
male_player_count = male_purchase_data["SN"].nunique()
male_player_percentage = 100 * male_player_count / total_players

# Create a dataframe with purchase data for female players
female_purchase_data = purchase_data.loc[purchase_data["Gender"] == "Female",:]

# Find the count and percentage of male players
female_player_count = female_purchase_data["SN"].nunique()
female_player_percentage = 100 * female_player_count / total_players

# Create a dataframe with purchase data for players with gender of "Other / Non-Disclosed"
other_gender_purchase_data = purchase_data.loc[purchase_data["Gender"] == "Other / Non-Disclosed",:]

# Find the count and percentage of male players
other_gender_player_count = other_gender_purchase_data["SN"].nunique()
other_gender_player_percentage = 100 * other_gender_player_count / total_players

# Create a summary table to display player gender demographics
gender_st = pd.DataFrame({"Total Count":[male_player_count,
                                         female_player_count,
                                         other_gender_player_count],
                          "Percentage of Players":[male_player_percentage,
                                                   female_player_percentage,
                                                   other_gender_player_percentage]},
                        index=["Male","Female","Other / Non-Disclosed"])

# Format the summary table to correctly display percentage for gender demographic data
gender_st["Percentage of Players"] = gender_st["Percentage of Players"].map("{:.2f}%".format)

# Display the summary table for gender demographic data
gender_st.head()

Total Count Percentage of Players
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [5]:
# Create a grouped data frame (grouped by gender)
gender_purchase_data = purchase_data.groupby("Gender")

# Generate a list of Purchase Count data by gender
gender_purchase_count_list = gender_purchase_data["Purchase ID"].count()

# Generate a list of Average Purchase Price by gender
gender_average_purchase_price_list = gender_purchase_data["Price"].mean()

# Generate a list of Total Purchase Value by gender
gender_purchase_value_list = gender_purchase_data["Price"].sum()

# Generate a list of Average Purchase Total per Person by gender
gender_avg_purchase_total = gender_purchase_data["Price"].sum() / gender_purchase_data["SN"].nunique()


# Create the summary data table to display gender purchase data
gender_purchase_st = pd.DataFrame({"Purchase Count":gender_purchase_count_list,
                                  "Average Purchase Price":gender_average_purchase_price_list,
                                  "Total Purchase Value":gender_purchase_value_list,
                                  "Average Purchase Total per Person":gender_avg_purchase_total})

# Format the summary data table with gender purchase data
gender_purchase_st["Average Purchase Price"] = gender_purchase_st["Average Purchase Price"].map("${:,.2f}".format)
gender_purchase_st["Total Purchase Value"] = gender_purchase_st["Total Purchase Value"].map("${:,.2f}".format)
gender_purchase_st["Average Purchase Total per Person"] = gender_purchase_st["Average Purchase Total per Person"].map("${:,.2f}".format)


# Display the summary data table for gender purchase data
gender_purchase_st

Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Average Purchase Total per Person  
Gender                                                                        
Female                             $361.94                             $4.47  
Male                             $1,967.64                             $4.07  
Other / Non-Disclosed               $50.19                             $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [6]:
# Create bins and labels to cut the player data by age
bins = [0,10,15,20,25,30,35,40,200]
column_labels = ["<10","10-14","15-19","20-24","25-29","30-34","35-39","40+"]

# Cut the data by age range and create a new dataframe with the additional column for age range
age_player_data = pd.cut(purchase_data["Age"],bins,right=False,labels=column_labels)
age_purchase_data = purchase_data
age_purchase_data["Age Range"] = age_player_data

# Group the data by Age Range and create a series based on the unique player counts in each age range
age_range_grouped = age_purchase_data.groupby("Age Range")

# Convert the series into a data frame so it can be displayed as a summary table
age_data_st = age_range_grouped["SN"].nunique().to_frame()
age_data_st.rename(columns = {"SN":"Total Count"}, inplace=True)
age_data_st

# Calculate the percentage for each age range based on total players
age_data_st["Percentage of Players"] = 100 * age_data_st["Total Count"] / total_players

# Format the percentages in the summary table
age_data_st["Percentage of Players"] = age_data_st["Percentage of Players"].map("{:.2f}%".format)
age_data_st

Total Count Percentage of Players
Age Range                                   
<10                 17                 2.95%
10-14               22                 3.82%
15-19              107                18.58%
20-24              258                44.79%
25-29               77                13.37%
30-34               52                 9.03%
35-39               31                 5.38%
40+                 12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [7]:
# Generate a list of Purchase Count by Age Range
age_purchase_count_list = age_range_grouped["Purchase ID"].count()

# Generate a list of Average Purchase Price by Age Range
age_average_purchase_price_list = age_range_grouped["Price"].mean()

# Generate a list of Total Purchase Value by Age Range
age_purchase_value_list = age_range_grouped["Price"].sum()
age_purchase_value_list

# Generate a list of Average Purchase Total per Person by Age Range
age_avg_purchase_total = age_range_grouped["Price"].sum() / age_range_grouped["SN"].nunique()


# Create the summary data table to display age range purchase data
age_purchase_st = pd.DataFrame({"Purchase Count":age_purchase_count_list,
                                  "Average Purchase Price":age_average_purchase_price_list,
                                  "Total Purchase Value":age_purchase_value_list,
                                  "Average Purchase Total per Person":age_avg_purchase_total})

# Format the summary data table with Age Range purchase data
age_purchase_st["Average Purchase Price"] = age_purchase_st["Average Purchase Price"].map("${:,.2f}".format)
age_purchase_st["Total Purchase Value"] = age_purchase_st["Total Purchase Value"].map("${:,.2f}".format)
age_purchase_st["Average Purchase Total per Person"] = age_purchase_st["Average Purchase Total per Person"].map("${:,.2f}".format)


# Display the summary data table for Age Range purchase data
age_purchase_st

Purchase Count Average Purchase Price Total Purchase Value  \
Age Range                                                               
<10                    23                  $3.35               $77.13   
10-14                  28                  $2.96               $82.78   
15-19                 136                  $3.04              $412.89   
20-24                 365                  $3.05            $1,114.06   
25-29                 101                  $2.90              $293.00   
30-34                  73                  $2.93              $214.00   
35-39                  41                  $3.60              $147.67   
40+                    13                  $2.94               $38.24   

          Average Purchase Total per Person  
Age Range                                    
<10                                   $4.54  
10-14                                 $3.76  
15-19                                 $3.86  
20-24                                 $4.32  
25-29                                 $3.81  
30-34                                 $4.12  
35-39                                 $4.76  
40+                                   $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [8]:
# Create a grouped data frame by player screen-name
screen_name_grouped = purchase_data.groupby("SN")

# Generate a list of the number of purchases each player made
player_purchase_count = screen_name_grouped["Purchase ID"].count()

# Generate a list of the average purchase price for each player
player_avg_purchase_price = screen_name_grouped["Price"].mean()

# Generate a list of the total purchase value of each player
player_total_purchase_value = screen_name_grouped["Price"].sum()

# Create a summary data table to display individual player purchase data
player_purchase_st = pd.DataFrame({"Purchase Count":player_purchase_count,
                                  "Average Purchase Price":player_avg_purchase_price,
                                  "Total Purchase Value":player_total_purchase_value})

# Sort the player purchase data by purchase count
sorted_purchase_st = player_purchase_st.sort_values(["Total Purchase Value"],ascending=False)

# Format the sorted player purchase data
sorted_purchase_st["Average Purchase Price"] = sorted_purchase_st["Average Purchase Price"].map("${:,.2f}".format)
sorted_purchase_st["Total Purchase Value"] = sorted_purchase_st["Total Purchase Value"].map("${:,.2f}".format)

# Display sorted player purchase data
sorted_purchase_st.head()

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, average item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [9]:
# Create a grouped data frame by item ID & item name
item_purchase_grouped = purchase_data.groupby(["Item ID","Item Name"])

# Generate a list of Purchase Count for each item
item_purchase_count = item_purchase_grouped["Purchase ID"].count()

# Generate a list of (average) Item Price for each item
item_price = item_purchase_grouped["Price"].mean()

# Generate a list of Total Purchase Value by item
item_total_purchase_value = item_purchase_grouped["Price"].sum()

# Create a summary data table for the item purchase data
item_purchase_st = pd.DataFrame({"Purchase Count":item_purchase_count,
                                "Item Price":item_price,
                                "Total Purchase Value":item_total_purchase_value})

# Sort the summary data table by Purchase Count
sorted_item_purchase_st = item_purchase_st.sort_values(["Purchase Count"],ascending=False)

# Format the summary data table for the item purchase data
sorted_item_purchase_st["Item Price"] = sorted_item_purchase_st["Item Price"].map("${:,.2f}".format)
sorted_item_purchase_st["Total Purchase Value"] = sorted_item_purchase_st["Total Purchase Value"].map("${:,.2f}".format)


# Display the summary data table for the item purchase data
sorted_item_purchase_st.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
132     Persuasion                                                 9   
108     Extraction, Quickblade Of Trembling Hands                  9   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
132     Persuasion                                        $3.22   
108     Extraction, Quickblade Of Trembling Hands         $3.53   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
132     Persuasion                                                 $28.99  
108     Extraction, Quickblade Of Trembling Hands                  $31.77

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [10]:
# Use the same grouped data frame and summary table from before
# Sort the summary data table by Total Purchase Value
value_sorted_item_purchase_st = item_purchase_st.sort_values(["Total Purchase Value"],ascending=False)

# Format the summary data table sorted by total purchase value
value_sorted_item_purchase_st["Item Price"] = value_sorted_item_purchase_st["Item Price"].map("${:,.2f}".format)
value_sorted_item_purchase_st["Total Purchase Value"] = value_sorted_item_purchase_st["Total Purchase Value"].map("${:,.2f}".format)


# Display the summary data table sorted by total purchase value
value_sorted_item_purchase_st.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

## Written Analysis

One of the trends I observed from this data is that while there five times more male players (and males have a higher total purchase value), players identifying as other, non-disclosed or female had a higher average total purchase per person, by over 50 cents on average. One way to generate more revenue could be to market the game more to non-male-identifying people.

Another observation is that while almost half of purchases were made by people in the 20-24 age-range (generating 46.81% of total purchase value), two age-range groups had higher average total purchase values per person (ages less than 10 and ages 35-39). Perhaps older players have either more income or less debt (which could explain why the age-range of 35-39 had a higher purchase on average per person), but that does not explain why the age range of "<10" had a higher average. We might consider that players less than 10 may require purchase power from parents (who may be in an older age-range, themselves). Having game content and marketing geared toward younger generation could generate more revenue, as long as that content and marketing takes parental approval/satisfaction into account.

Finally, an observation about item-specific purchase data is that items that were popular or had a higher total purchase value had a much higher average item price (between $4.23 – $4.61 for the top five items) than the overall average item price ($3.05). We might interpret this to mean that item pricing is not a barrier for purchases, since more expensive items are popular (and slightly less-popular items are still generating high total purchase values).